gensim/similarities/levenshtein.py

```diff
@@ -8,146 +8,101 @@
     alpha : float, optional
-        The multiplicative factor alpha defined by Charlet and Damnati (2017).
+        The multiplicative factor alpha defined by [charletetal17]_.
     beta : float, optional
-        The exponential factor beta defined by Charlet and Damnati (2017).
-    threshold : float, optional
-        Only terms more similar than `threshold` are considered when retrieving
-        the most similar terms for a given term.
+        The exponential factor beta defined by [charletetal17]_.
+    max_distance : int, optional
+        The maximum Levenshtein distance of the most similar terms.
+        Keeping this value below 3 has a significant impact on the
+        retrieval performance. Default is 1.
+
+    Attributes
+    ----------
+    dictionary : :class:`~gensim.corpora.dictionary.Dictionary`
+        A dictionary that specifies the considered terms.
+    alpha : float
+        The multiplicative factor alpha defined by [charletetal17]_.
+    beta : float
+        The exponential factor beta defined by [charletetal17]_.
+    index : :class:`lexpy.dawg.DAWG`
+        The DAWG nearest-neighbor search index.
+    max_distance : int
+        The maximum Levenshtein distance of the most similar terms.
+        Keeping this value below 3 has a significant impact on the
+        retrieval performance.
 
     See Also
     --------
-    :func:`gensim.similarities.levenshtein.levsim`
-        The Levenshtein similarity.
+    :class:`~gensim.similarities.termsim.WordEmbeddingSimilarityIndex`
+        Retrieve most similar terms for a given term using the cosine similarity over word
+        embeddings.
     :class:`~gensim.similarities.termsim.SparseTermSimilarityMatrix`
         Build a term similarity matrix and compute the Soft Cosine Measure.
 
+    References
+    ----------
+    The Levenshtein similarity in the context of term similarity was defined
+    by [charletetal17]_.
+
+    .. [charletetal17] Delphine Charlet and Geraldine Damnati, "SimBow at SemEval-2017 Task 3:
+       Soft-Cosine Semantic Similarity between Questions for Community Question Answering", 2017,
+       https://www.aclweb.org/anthology/S17-2051/.
+
     """
-    def __init__(self, dictionary, alpha=1.8, beta=5.0, threshold=0.0):
+    def __init__(self, dictionary, alpha=1.8, beta=5.0, max_distance=1):
         self.dictionary = dictionary
         self.alpha = alpha
         self.beta = beta
-        self.threshold = threshold
+        self.max_distance = max_distance
+
+        from lexpy.dawg import DAWG
+
+        self.index = DAWG()
+        terms = sorted(self.dictionary.values())
+        self.index.add_all(terms)
+        self.index.reduce()
+
         super(LevenshteinSimilarityIndex, self).__init__()
 
+    def _levsim(self, t1, t2):
+        from Levenshtein import distance
+
+        max_lengths = max(len(t1), len(t2)) or 1
+        similarity = self.alpha * (1.0 - distance(t1, t2) * 1.0 / max_lengths)**self.beta
```

Comment:
Does `lexpy` have an option to return the distances too, along with the terms? For sure it must know them.
It seems wasteful to recompute again here.

Reply:
It does not seem that way, see https://github.com/aosingh/lexpy/blob/master/lexpy/_base/automata.py#L261. However, the performance impact of computing the Levenshtein distances for a reasonably small `topn` is tiny, I measured this.

Reply:
OK. Not affecting performance – just weird API.